### PR TITLE
ThreatModels: Add `stdin` kind

### PR DESCRIPTION
**Pull request by RasmusWL**
As discussed internally, none of the current local subgroups precisely captures stdin, so although it's much like both commandargs and file, a separate kind seems better.

I've ported the obvious models from C# and Java.

### CODE REVIEW

**Comment by RasmusWL** [comment]
I don't think this warrants a change-note, but let me know if you think otherwise :+1:

**Comment by michaelnebel** [comment]
> I don't think this warrants a change-note, but let me know if you think otherwise 👍

Thank you @RasmusWL !

Just to be on the safe side, I think we should add a change note. At least for Java we are "moving" `System.in` into a new threat model group (for C# it seems less important with a change note as we are just making the modelling more fine grained).
Some tests that uses specific threat model groups might need an update (the threat model configuration is located in the `.ext.yml` file next to the test file).

File: shared/threat-models/ext/threat-model-grouping.model.yml [hunk 1]
```diff
@@ -15,6 +15,7 @@ extensions:
       - ["database", "local"]
       - ["commandargs", "local"]
       - ["environment", "local"]
+      - ["stdin", "local"]
```
Comment (egregius313) [inline shared/threat-models/ext/threat-model-grouping.model.yml:18]: Overall I think this is a good decision.

But I'm debating whether it makes more sense as a completely separate threat model kind versus a sub-kind of `file` (my understanding of how `threatModelEnabled` works is that it supports sub-kinds).

Either way I agree with @michaelnebel that this should have a change note.

**Comment by owen-mc** [comment]
I agree it would be good to have a change note so that users who read the change notes are aware of the new threat model kind. You can probably copy some of the test from [this change note](https://github.com/github/codeql/pull/16760/files#diff-e4d8dae8f74704a02f34a0b9ddd094dd24f768eee4a75cffad3a150d352eb140) from a PR where I added a new kind (but in a slightly different context, where I was removing some sources from the default and letting users know how to turn them back on).

I find @egregius313 's suggestion of making it a sub-model of `file` interesting. But I don't know enough about appsec to say what makes the most sense.

**Review by owen-mc — CHANGES_REQUESTED** [review]
Please also update the documentation listing different threat model kinds [here](https://github.com/github/codeql/blob/main/docs/codeql/reusables/threat-model-description.rst).

**Comment by RasmusWL** [comment]
I find it odd that I had to change MaD numbering in 224fe12, and I'm worried it's an indicator of an underlying problem (why would the numbering change for other parts of the modeling, when I'm only _changing_ the threat-model of some stuff??) -- but I don't have enough expertise to debug this further. I did evaluate `sourceModel` locally on a random C# DB and confirmed that the 3 stdin models did not change madId after the threat-model was changed to stdin.

**Comment by michaelnebel** [comment]
> I find it odd that I had to change MaD numbering in [224fe12](https://github.com/github/codeql/commit/224fe12f38d68797a97cf2ab27f2fb09d9bd167b), and I'm worried it's an indicator of an underlying problem (why would the numbering change for other parts of the modeling, when I'm only _changing_ the threat-model of some stuff??) -- but I don't have enough expertise to debug this further. I did evaluate `sourceModel` locally on a random C# DB and confirmed that the 3 stdin models did not change madId after the threat-model was changed to stdin.

I think the reason you are seeing a "shift" in the MaD IDs is because they are based directly on the `QlBuiltins::ExtensionId`, and the PR adds on extension row (the `stdin`) threat model group.
In any case you need to rebase the PR: I re-wrote all the C# tests yesterday to use the new post processing logic, which should increase test stability as the extension id (built-in) is no longer used directly in the expected test output.

**Comment by RasmusWL** [comment]
> the PR adds an extension row (the `stdin`) threat model group.

Ah yes, that explains the difference :+1: I've rebased the PR :+1:

**Comment by michaelnebel** [comment]
@RasmusWL : Excellent! Thank you!

**Review by owen-mc — APPROVED** [review]

**Comment by RasmusWL** [comment]
The `C# Integration tests MacOS` seems like a flaky failure that can be ignored, so should be ready to be merged now :+1:

**Review by owen-mc — APPROVED** [review]

**Review by michaelnebel — APPROVED** [review]
Looks great!

**Pull request merged**